HDDS-7121. Support namespace summaries (du, dist & counts) for legacy FS buckets

xBis7 opened the pull request:
## What changes were proposed in this pull request?

In the previous PR [HDDS-5504](https://github.com/apache/ozone/pull/3309), we refactored `NSSummaryEndpoint` and `NSSummaryTask` so that we can extract the bucket specific code to new separated classes. So far, Recon has only been supporting FSO buckets. All the existing bucket related code exists in `handlers/FSOBucketHandler` and `tasks/NSSummaryTaskWithFSO`. In order to add support for Legacy buckets, we created two new classes `handlers/LegacyBucketHandler` and `tasks/NSSummaryTaskWithLegacy`.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-7121

## How was this patch tested?

This patch was tested with new unit tests. We added two new test classes, `TestNSSummaryEndpointWithLegacy` and `TestNSSummaryTaskWithLegacy`. These two classes are as similar as possible to the corresponding FSO classes. We tried to keep the setups exactly the same with `TestNSSummaryEndpointWithFSO` and `TestNSSummaryTaskWithFSO` as if they were handling FSO related code.

GeorgeJahad commented:
So there is a problem with clearing the NSSummary table during reprocess().

Both the fso and legacy tasks write to the nssummary table. The reprocess() method for FSO clears the nssummary table [here](https://github.com/apache/ozone/blob/39c0c25c07e359a1e5f0968a2143db1dc9da3c19/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithFSO.java#L176) The Legacy task doesn't clear the table.

But both tasks run in separate threads in [parallel](https://github.com/apache/ozone/blob/39c0c25c07e359a1e5f0968a2143db1dc9da3c19/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ReconTaskControllerImpl.java#L172-L179)  There is no guarantee that FSO will run first.

In addition failed tasks get rerun [here](https://github.com/apache/ozone/blob/39c0c25c07e359a1e5f0968a2143db1dc9da3c19/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ReconTaskControllerImpl.java#L120-L131)  If one fails and the other doesn't the delete won't work correctly.

GeorgeJahad commented:
Also because the two tasks are writing to the same table from separate threads, it would be could to clarify why no synchronization is needed.

dombizita commented:
Hi @xBis7 and @GeorgeJahad, thanks for the patch and the problem description. I believe if we already separated the NSSummary handling for legacy and FSO buckets ([here](https://github.com/apache/ozone/pull/3309)) we can maintain two separate NSSummary tables, one for each bucket layout. This would help with the table clear and if one of them fails and there is a retry. Let me know what you think of this idea. cc @smengcl

xBis7 commented:
@dombizita @smengcl Another solution to this, would be to have a common NSSummary table for both bucket layouts. Instead of calling both tasks, we could add `process` and `reprocess` methods to `NSSummaryTask` and call that class in place of the subclasses. `NSSummaryTask.process` will call both `NSSummaryTaskWithFSO.process` and `NSSummaryTaskWithLegacy.process` and we will do the same for `reprocess`. `NSSummaryTask` will also manage common setups or operations for NSSummary table, such as clearing it. What do you think about this approach?

dombizita commented:
If we have two separate tables for `NSSummary`s based on the bucket layout (legacy and FSO) we can iterate through both tables right after each other if there is a task that needs information from both types. I think this approach would be nice to avoid issues as we already separated the classes based on the bucket layout.
But as I looked into it for both bucket types we would maintain a table with the same columns, as we add long value (ID) and a `NSSummary` in both cases. Based on this it would be better to avoid having two same tables.
Your approach sounds good to me, if I understand correctly the `NSSummaryTask.process` would call eg. first `NSSummaryTaskWithFSO.process` and after the `NSSummaryTaskWithLegacy.process`, right?

xBis7 commented:
@dombizita `NSSummary` isn't bucket specific. We would have an issue if we were gathering info for volumes but we are dealing with buckets. 

> NSSummaryTask.process would call eg. first NSSummaryTaskWithFSO.process and after the NSSummaryTaskWithLegacy.process, right?

Yes, you are right. This way we can control the order by which we are calling any bucket specific task. `NSSummaryTask.process` will return `new ImmutablePair<>(getTaskName(), true)` after both bucket tasks have finished running and `getTaskName()` will be `NSSummaryTask`. Since `NSSummary` isn't specific to bucket layout and the entries for different layouts won't have an attribute that differentiates them, it seems more clean to have one table instead of two tables with entries of the same type.

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/LegacyBucketHandler.java:
```diff
@@ -0,0 +1,317 @@
+
+    try {
+      dirObjectId = dirInfo.getObjectID();
+    } catch (NullPointerException e) {
```
why is the exception being swallowed?  it will cause the bucketid to be returned.

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,266 @@
+
+      try {
+        keyInfo.setParentObjectID(parentKeyInfo.getObjectID());
+      } catch (NullPointerException e) {
```
why is the npe being swallowed?

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,266 @@
+
+      try {
+        keyInfo.setParentObjectID(parentBucketInfo.getObjectID());
+      } catch (NullPointerException e) {
```
why is the npe being swallowed?

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/BucketHandler.java:
```diff
@@ -183,8 +183,17 @@ public static BucketHandler getBucketHandler(
                 OmBucketInfo bucketInfo) throws IOException {
 
-    return new FSOBucketHandler(reconNamespaceSummaryManager,
+    if (bucketInfo.getBucketLayout()
```
This version of getBucketHandler will throw an NPE if bucketInfo is null.  The other version just returns [null](https://github.com/apache/ozone/blob/39c0c25c07e359a1e5f0968a2143db1dc9da3c19/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/BucketHandler.java#L210).  They should do the same thing, probably both throw.

xBis7 replied:
We need the other version to return null, so that we know that the BucketHandler doesn't exist and later return `EntityType.UNKNOWN`. See [here](https://github.com/xBis7/ozone/blob/HDDS-7121/hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/EntityHandler.java#L149-L158).

GeorgeJahad replied:
It's usually good practice to keep the methods consistent.  If the null check is required, it would be better to put it in the one that is called by this one, so they both have it.

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/DirectoryEntityHandler.java:
```diff
@@ -89,7 +91,15 @@ public DUResponse getDuResponse(
+        subdirName = subdirFileName.toString();
+      } else {
+        throw new NullPointerException("Subdirectory file name is null.");
+      }
```
is this change to remove the trailing slash?  why not: OzoneFSUtils.removeTrailingSlashIfNeeded()

xBis7 replied:
My mistake, I updated the comment to be more descriptive of the issue. Eg. the key name is `dir1/dir2`, then we just need to add the subdir to the path, we need `dir2`. Otherwise, we will end up having `dir1` twice while this is an iterative process going over all directories and adding them one by one.

GeorgeJahad replied:
So then why test for null and then do an explicit throw?  won't the throw happen even if you don't check for it?

xBis7 replied:
I added the null check to get rid of a `findbugs` error but I also didn't want to consume the NPT exception. I can change the exception type to `IOException`.

```
M D NP: Possible null pointer dereference in org.apache.hadoop.ozone.recon.api.handlers.DirectoryEntityHandler.getDuResponse(boolean, boolean) due to return value of called method  Dereferenced at DirectoryEntityHandler.java:[line 102]
```

xBis7 commented:
@GeorgeJahad I've made the changes based on your recommendations.

xBis7 commented:
@dombizita I've made the changes so that `NSSummaryTask` gets called in place of `NSSummaryTaskWithFSO` and `NSSummaryTaskWithLegacy`, like we discussed. Let me know how it looks. I've also modified the tests to go only over each subclass and not `NSSummaryTask`. If we decide to move forward with this approach I will cleanup the code and add a new class `TestNSSummaryTask` that will test the case of having buckets of different layouts.

dombizita reviewed (commented):
Thanks @xBis7 for updating your patch, this approach of processing and reprocessing looks good to me, I left a few inline comments.

xBis7 commented on hadoop-ozone/dist/src/main/smoketest/recon/recon-nssummary.robot:
```diff
@@ -14,7 +14,7 @@
 
 *** Settings ***
-Documentation       Smoke test for Recon Namespace Summary Endpoint for FSO buckets.
+Documentation       Smoke test for Recon Namespace Summary Endpoint for FSO and Legacy buckets.
```
@dombizita Apart from the Namespace Summary for a directory there is no difference between FSO and LEGACY. In order to avoid duplication, we added the `BUCKET_LAYOUT` as a parameter and gave it the default value of FSO. The idea is to add an extra command to `run.sh` to also run the test with a parameter `-v BUCKET_LAYOUT:LEGACY`.

xBis7 commented on hadoop-ozone/dist/src/main/smoketest/recon/recon-nssummary.robot:
```diff
@@ -131,7 +132,8 @@ Check Recon Namespace Summary Key
 
 Check Recon Namespace Summary Directory
-    Wait For Summary      ${SUMMARY_URL}?path=/${VOLUME}/${BUCKET}/dir1/dir2   DIRECTORY
+    Run Keyword If    '${BUCKET_LAYOUT}' == 'LEGACY'                    Wait For Summary      ${SUMMARY_URL}?path=/${VOLUME}/${BUCKET}/dir1/dir2/   DIRECTORY
```
After running the test with the parameter `-v BUCKET_LAYOUT:LEGACY`, here it will execute the part for LEGACY.

xBis7 replied:
The trailing slash is necessary. Without it, we get `PATH_NOT_FOUND` and entity `UNKNOWN`.

> the command after this line 135 will only be executed when '${BUCKET_LAYOUT}' == 'LEGACY'.
If bucket layout is FSO, it will run the command line below instead, with the subtle difference of the trailing slash in the directory path.

You are right. That's how it works.

We give `${BUCKET_LAYOUT}` a default value of FSO and [here](https://github.com/xBis7/ozone/blob/HDDS-7121/hadoop-ozone/dist/src/main/compose/ozone-legacy/test.sh#L31) we execute the smoketest with value LEGACY.

smengcl replied:
Thanks @xBis7 for the explanation. The link would now be [this](https://github.com/xBis7/ozone/blob/HDDS-7121/hadoop-ozone/dist/src/main/compose/ozone-legacy-bucket/test.sh#L31) since the dir is renamed.

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,255 @@
+    return true;
+  }
+
+  private void setKeyParentID(OmKeyInfo keyInfo) throws IOException {
```
Add a comment explaining that legacy objects don't have the parentid set, so we have to do it explicitly using this method.

GeorgeJahad commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/tasks/TestNSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,734 @@
+        VOL,
+        FILE_ONE,
+        KEY_ONE_OBJECT_ID,
+        BUCKET_ONE_OBJECT_ID,
```
The parent object id for legacy records is 0, (which is why we added the setKeyParentId() method.)  Let's set them back to 0 in the test code as well, just to be sure that setKeyParentId() is really working.

GeorgeJahad commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestNSSummaryEndpointWithLegacy.java:
```diff
@@ -0,0 +1,1286 @@
+          VOL,
+          DIR_ONE,
+          DIR_ONE_OBJECT_ID,
+          BUCKET_ONE_OBJECT_ID,
```
The parent object id for legacy records is 0, (which is why we added the setKeyParentId() method.) Let's set them back to 0 in the test code as well, just to be sure that setKeyParentId() is really working.

GeorgeJahad commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/api/TestNSSummaryEndpointWithLegacy.java:
```diff
@@ -0,0 +1,1286 @@
+        VOL,
+        FILE_ONE,
+        KEY_ONE_OBJECT_ID,
+        BUCKET_ONE_OBJECT_ID,
```
The parent object id for legacy records is 0, (which is why we added the setKeyParentId() method.) Let's set them back to 0 in the test code as well, just to be sure that setKeyParentId() is really working.

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTask.java:
```diff
@@ -45,165 +41,71 @@
-      LOG.error("The namespace table is not correctly populated.");
-      return;
+  @Override
+  public Pair<String, Boolean> process(OMUpdateEventBatch events) {
```
Now that the two tasks have been merged we'll need to add a simple test.  one fso bucket and one legacy bucket, each with one key.

GeorgeJahad replied:
Also, please sure the test coverage results

xBis7 replied:
okay, I suppose this will be under a new class, `TestNSSummaryTask`.

GeorgeJahad replied:
So TestNSSummaryTask is basically a copy of all the other tests.  As I mentioned above, instead of unnecessarily duplicating all those tests, it would be better to "add a simple test. one fso bucket and one legacy bucket, each with one key."

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTask.java:
```diff
@@ -45,165 +41,71 @@
-    if (nsSummary == null) {
-      LOG.error("The namespace table is not correctly populated.");
-      return;
+    success = nsSummaryTaskWithFSO.reprocessWithFSO(omMetadataManager);
```
same as above

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,255 @@
+          omMetadataManager.getKeyTable(BUCKET_LAYOUT);
+
+      try (TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>>
+          keyTableIter = keyTable.iterator()) {
```
It just occurred to me that since both the legacy and the obs keys are both in the keytable, and this task is just for legacy keys, it will have to explicitly skip obs keys.

xBis7 replied:
The only way I see to do that, is for each OmKeyInfo, get the bucketName and using that, get the bucket info from the bucket table and then check the layout.

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskUtils.java:
```diff
@@ -0,0 +1,197 @@
+ * related to DB operations so that they can commonly be
+ * used in NSSummaryTaskWithFSO and NSSummaryTaskWithLegacy.
+ */
+public class NSSummaryTaskUtils {
```
NSSummaryTaskUtils is not a great name for a parent class.  How about NSSummaryDbEventHandler?

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -185,6 +185,21 @@ public boolean reprocessWithLegacy(OMMetadataManager omMetadataManager) {
+              .getBucketTable().getSkipCache(bucketDBKey);
+
+          if (omBucketInfo.getBucketLayout().isObjectStore()) {
+            continue;
```
There should be a simple test that confirms this works.

GeorgeJahad replied:
Also, shouldn't this be in process() as well?

GeorgeJahad commented:
I just tried running: "ozone admin namespace summary /vol1"

It gets a warning:
```
[Warning] Namespace CLI is only designed for FSO mode.
Bucket being accessed must be of type FILE_SYSTEM_OPTIMIZED bucket layout.
```
It should be updated to something like this:
```
[Warning] Namespace CLI is only designed for FSO buckets or LEGACY buckets with "ozone.om.enable.filesystem.paths" set to true.
```

GeorgeJahad commented on hadoop-ozone/dist/src/main/compose/ozone-legacy/docker-config:
```diff
@@ -0,0 +1,51 @@
+CORE-SITE.XML_fs.trash.interval=1
+
+OZONE-SITE.XML_ozone.om.address=om
+OZONE-SITE.XML_ozone.om.enable.filesystem.paths=true
```
Should probably add:
OZONE-SITE.XML_ozone.default.bucket.layout=LEGACY

just in case the default changes in the future

GeorgeJahad commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/tasks/TestNSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,735 @@
+        VOL,
+        FILE_ONE,
+        KEY_ONE_OBJECT_ID,
+        PARENT_OBJECT_ID_ZERO,
```
Please do the same for populateOMDB() and setUpMultiBlockReplicatedKeys() in TestNSSummaryEndpointWithLegacy.java

GeorgeJahad commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/BucketLayout.java:
```diff
@@ -70,6 +70,10 @@ public boolean isLegacy() {
   }
 
+  public boolean isObjectStore() {
+    return this.equals(OBJECT_STORE);
```
A legacy bucket with "ozone.om.enable.filesystem.paths" set to false is also an object store bucket.  so test for that here as well.

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,276 @@
+              keyInfo.getBucketName(), parentKeyName);
+      OmKeyInfo parentKeyInfo = getReconOMMetadataManager()
+          .getKeyTable(BUCKET_LAYOUT)
+          .get(fullParentKeyName);
```
getSkipCache()

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,276 @@
+      String bucketKey = getReconOMMetadataManager()
+          .getBucketKey(keyInfo.getVolumeName(), keyInfo.getBucketName());
+      OmBucketInfo parentBucketInfo =
+          getReconOMMetadataManager().getBucketTable().get(bucketKey);
```
getSkipCache()

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,309 @@
+        OmBucketInfo omBucketInfo = getReconOMMetadataManager()
+            .getBucketTable().getSkipCache(bucketDBKey);
+        // True if FileSystemPaths enabled
+        boolean enableFileSystemPaths =
```
This can be set at the top of the method, outside of the while loop.

GeorgeJahad commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,309 @@
+          OmBucketInfo omBucketInfo = omMetadataManager
+              .getBucketTable().getSkipCache(bucketDBKey);
+          // True if FileSystemPaths enabled
+          boolean enableFileSystemPaths =
```
same as above

GeorgeJahad commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/tasks/TestNSSummaryTask.java:
```diff
@@ -0,0 +1,491 @@
+
+  /**
+   * Create a new OM Metadata manager instance with one user, one vol, and two
+   * buckets. Bucket1 will have FSO layout and bucket2 will have Legacy layout.
```
mention bucket 3 here as well

GeorgeJahad commented:
LGTM

xBis7 commented:
@dombizita @smengcl Could you take a look at this patch?

smengcl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,309 @@
+    Iterator<OMDBUpdateEvent> eventIterator = events.getIterator();
+    Map<Long, NSSummary> nsSummaryMap = new HashMap<>();
+    // true if FileSystemPaths enabled
+    boolean enableFileSystemPaths =
+        ozoneConfiguration.get(OMConfigKeys
+            .OZONE_OM_ENABLE_FILESYSTEM_PATHS).equals("true");
```
but here we would want to read the config key only once in the constructor and store it to a local variable. then use that local variable here.

also note: config value case sensitivity.

smengcl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,309 @@
+  public boolean reprocessWithLegacy(OMMetadataManager omMetadataManager) {
+    Map<Long, NSSummary> nsSummaryMap = new HashMap<>();
+    // true if FileSystemPaths enabled
+    boolean enableFileSystemPaths =
+        ozoneConfiguration.get(OMConfigKeys
+            .OZONE_OM_ENABLE_FILESYSTEM_PATHS).equals("true");
```
here as well

smengcl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTaskWithLegacy.java:
```diff
@@ -0,0 +1,309 @@
+  private void setKeyParentID(OmKeyInfo keyInfo) throws IOException {
+    String[] keyPath = keyInfo.getKeyName().split(OM_KEY_PREFIX);
+
+    //if (keyPath > 1) there is one or more directories
```
nit: rephrase this comment (make this line of comment more readable)

smengcl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTask.java:
```diff
@@ -44,166 +49,97 @@
+    success = nsSummaryTaskWithFSO.processWithFSO(events);
+    if (success) {
+      success = nsSummaryTaskWithLegacy.processWithLegacy(events);
     }
```
more logging
```suggestion
    } else {
        LOG.error("processWithFSO failed");
    }
```

smengcl commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/NSSummaryTask.java:
```diff
@@ -44,166 +49,97 @@
+
+    List<Future<Boolean>> results;
+    ExecutorService executorService = Executors
+        .newFixedThreadPool(2);
+    try {
+      results = executorService.invokeAll(tasks);
```
Does this imply `reprocessWithFSO` and `reprocessWithLegacy` will be run in parallel?

xBis7 replied:
I'm not sure if they will run in parallel but they will run asynchronously on different threads. We did this to optimize reprocess since it takes much longer than process to finish running.

smengcl replied:
I see. As long as this won't cause similar issue we discussed much earlier (one thread clearing NSSummaryTable while another is writing to it) it should be fine. I see `clearNSSummaryTable()` being called right before invoke here.

xBis7 replied:
There should be no issues. We clear the NSSummary table only once and then we invoke the tasks and let them run asynchronously.

smengcl replied:
Sounds good. Thanks!

smengcl reviewed (commented):
Thanks @xBis7 for the improvements and test additions. I have some minor comments inline.

xBis7 commented on hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/handlers/BucketHandler.java:
```diff
@@ -183,8 +185,22 @@ public static BucketHandler getBucketHandler(
+        return new LegacyBucketHandler(reconNamespaceSummaryManager,
+            omMetadataManager, reconSCM, bucketInfo);
+      } else {
+        throw new RuntimeException("Unsupported bucket layout.");
```
> Could it crash the Recon thread / Recon process?

Probably... This exception will come from hitting an endpoint in `NSSummaryEndpoint`. We could wrap it with a try catch and return a `Response.serverError()` in the catch block. What do you think?

smengcl replied:
Yup let's return an error instead of potentially crashing Recon

xBis7 replied:
Log an error instead of throwing the exception or catch the exception and then log an error?

smengcl replied:
Okay let's just do `LOG.error` then. No need to throw unchecked exception here.

xBis7 commented:
@smengcl thanks for the review. I 've addressed all your comments except [this one](https://github.com/apache/ozone/pull/3746#discussion_r998884262). Do you want to do something about the RunTimeException or do you want to leave prolong and crash the thread?

smengcl commented on hadoop-ozone/recon/src/test/java/org/apache/hadoop/ozone/recon/OMMetadataManagerTestUtils.java:
```diff
@@ -256,6 +255,28 @@ public static void writeKeyToOm(OMMetadataManager omMetadataManager,
+   * Write a directory as key on OM instance.
+   * We don't need to set size.
+   * @throws IOException
+   */
```
Sounds good. Thanks!

smengcl commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/nssummary/NSSummaryAdmin.java:
```diff
@@ -108,6 +110,60 @@ public boolean isFileSystemOptimizedBucket(String path) throws IOException {
+    OFSPath ofsPath = new OFSPath(path);
+
+    boolean enableFileSystemPaths = getOzoneConfig()
+        .get(OMConfigKeys.OZONE_OM_ENABLE_FILESYSTEM_PATHS).equals("true");
```
Thanks for the change @xBis7 !

Though my bad I just realized we should probably straight use `.getBoolean()` with the default value, like this:

https://github.com/apache/ozone/blob/c0de6bcdebee6a4271fc73ea6542902e0eeee28f/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java#L207-L209

smengcl commented on hadoop-ozone/dist/src/main/compose/ozone-legacy/README.md:
```diff
@@ -0,0 +1,21 @@
+<!---
```
Thanks!

xBis7 commented:
@smengcl I pushed the latest changes for getting `FileSystemPaths` value from configuration.

smengcl reviewed (approved):
Latest changes lgtm. Thanks @xBis7 . Pending CI.

smengcl commented:
Merged. Thanks @xBis7 for the work! Thanks @GeorgeJahad and @dombizita for the review.
